### PR TITLE
feat(ci): add PR title check workflow

### PR DESCRIPTION
**Pull request by rjan90**
## Related Issues
Closes: #12148

## Proposed Changes
- A GitHub Actions workflow to automatically check and enforce a consistent format for PR titles
- Started work on `CONTRIBUTING.md` doc that consolidates contributing expectations in one place. https://github.com/filecoin-project/lotus/issues/12360

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] Update CHANGELOG.md or signal that this change does not need it.
  - If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
  - If the change does not require a CHANGELOG.md entry, do one of the following:
    - Add `[skip changelog]` to the PR title
    - Add the label `skip/changelog` to the PR
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by rjan90** [comment]
> Create a PR with an invalid title format and verify that it's blocked

Confirm that creating a PR with a invalid title format is being requested for change by the GA bot, [as per this comment.](https://github.com/filecoin-project/lotus/pull/12340#pullrequestreview-2219004254)

**Comment by rjan90** [comment]
>  Edit a PR title from invalid to valid format and confirm the check passes
>  Confirm that "Changes requested" gets removed once updated to the correct format.

After updating my title to valid format, I can confirm that the check passes, and also that the "Changes requested" by the bot gets dismissed:
Screenshot 2024-08-05 at 15 30 12

File: .github/workflows/pr-title-check.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    name: Check PR Title
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check PR Title and Manage Review
+        uses: actions/github-script@v6
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            const title = context.payload.pull_request.title;
+            const pattern = /^(feat|fix|build|chore|ci|docs|perf|refactor|revert|style|test):\s\w+:\s.+/;
```
Comment (rvagg) [inline .github/workflows/pr-title-check.yml:21]:
```suggestion
            const pattern = /^(feat|fix|build|chore|ci|docs|perf|refactor|revert|style|test):\s\w+:\s[a-z].+/;
```

"Subject: start with lowercase"

File: .github/workflows/pr-title-check.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    name: Check PR Title
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check PR Title and Manage Review
+        uses: actions/github-script@v6
+        with:
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            const title = context.payload.pull_request.title;
+            const pattern = /^(feat|fix|build|chore|ci|docs|perf|refactor|revert|style|test):\s\w+:\s[a-z].+/;
```
Comment (BigLep) [inline .github/workflows/pr-title-check.yml:21]: (just cluing into this now while thinking about it fresh) I'm curious why we deviate from conventional commit style: https://www.conventionalcommits.org/en/v1.0.0/ ?

We do:

``feat: api: description``

Conventional commit has the "scope" with "()":

``feat(api): description``

---
Also, nit, but maybe alphabetize the "types"?

``build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test``

File: .github/workflows/pr-title-check.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            
+            if (!pattern.test(title)) {
+              await github.rest.pulls.createReview({
+                ...context.repo,
+                pull_number: context.payload.pull_request.number,
+                body: 'Please update the PR title to match the required format: `<type>: <area>: <subject>`\n\nTypes: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test, chore\nArea: e.g. api, chain, state, mempool, multisig, networking, paych, proving, sealing, wallet, deps\nSubject: start with lowercase',
```
Comment (BigLep) [inline .github/workflows/pr-title-check.yml:27]:
```suggestion
                body: 'Please update the PR title to match the required format: `<type>: <scope>: <subject>`\n\nTypes: build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test\nScope: e.g. api, chain, deps, state, mempool, multisig, networking, paych, proving, sealing, wallet\nSubject: start with lowercase',
```

Alphabetized
Used conventional commit "scope" terminology

Further updates if we do the ``type(scope): summary`` conventional commit syntax.

**Review by BigLep — COMMENTED** [review]
A few thoughts:
1. (raised inline) Any reason we're not just following https://www.conventionalcommits.org/en/v1.0.0/  style?
2. Do we expect to want to roll this out to other FilOz-maintained repos?  I assume IPDX has ideas on how this could be done (asked [here](https://filecoinproject.slack.com/archives/C03KLC57LKB/p1723059192656669)).
3. What are the thoughts on using a 3rd party action?  I realize there's not a lot of code here so it's not a big gain, but others have solved this problem.  Examples:
   - https://github.com/marketplace/actions/pr-title-checker
   - https://github.com/marketplace/actions/pr-title-verify
   - https://github.com/Slashgear/action-check-pr-title
   My impetus for raising this is the value for keeping it concise probably goes up if we do want to copy this across multiple repos.

**Comment by rvagg** [comment]
fwiw I prefer conventional commit style and use it everywhere else, this is the repo that stands out as different in the world of repos I contribute to; but change can be hard for people that have got used to it

One option is to do both for a while and have a migration period.

```
^(feat|fix|build|chore|ci|docs|perf|refactor|revert|style|test)(\(\w+\))?:\s\w+:\s[a-z].+|^(feat|fix|build|chore|ci|docs|perf|refactor|revert|style|test)(\(\w+\))?:\s[a-z].+
```

**Comment by BigLep** [comment]
> migration period idea

Seems good/reasonable to me.  We could still update our PR template text to steer at the conventional commit style.

Also, I'm realizing we'll need to update the regex to account for `[skip changelog]`

**Comment by rjan90** [comment]
1. I agree that moving to the conventional commit style is preferred, and we should have a migration period where both styles are acceptable.

2. I think it would be beneficial to roll this out to the repositories with higher activity and importance (builtin-actors, filecoin-ffi, ref-fvm, go-state-types, go-f3). For many of the smaller repositories with infrequent or less activity, I'm not sure if the benefits of such a title check outweigh the additional work (and potentially cost). However, I'm open-minded if there is an easy way to roll this out to all repositories maintained by FilOz.

3. I will look into those 3rd party actions. Perhaps we can switch to one of these after we have completed the migration period here in the Lotus repo.

I will adjust the PR accordignly

**Comment by rjan90** [comment]
In https://github.com/filecoin-project/lotus/pull/12340/commits/2cf80e7ecdfc8c4c89d6d654064e760f28f2cb24, I have changed:

- Allow optional "[skip changelog]" prefix
- Both `<type>: <scope>: <subject>` and `<type>(<scope>): <subject>` formats
- Use conventional commit `<scope>` instead of `<area>`
- Alphabetize <type> and <scope>

**Comment by rjan90** [comment]
While we are discussing conventions, referring to: [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).

> The commit contains the following structural elements, to communicate intent to the consumers of your library:
> .............
> 3. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a breaking API change (correlating with [MAJOR](http://semver.org/#summary) in Semantic Versioning). A BREAKING CHANGE can be part of commits of any type.

While this convention is primarily for commits (which Rod also has advocated for: https://github.com/filecoin-project/lotus/pull/12238#issuecomment-2229687995), it would be valuable to extend it to PR titles as well, particularly as we default to "squash and merge". By appending a ! after the <type> in the PR title, we can clearly signal that the change is breaking.

**Review by BigLep — APPROVED** [review]
I'm good with this getting merged once feedback incorporated.

I created another PR on top of your PR that starts extracting out some of these contributing conventions as part of https://github.com/filecoin-project/lotus/issues/12360

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -12,10 +12,10 @@
 Before you mark the PR ready for review, please make sure that:
 
 - [ ] Commits have a clear commit message.
-- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
+- [ ] PR title is in the form of of `<type> (<scope>): <subject>`
```
[the remark this answers is not in the document]
Comment (BigLep) [reply .github/pull_request_template.md:15]: I didn't create a commit.  Here's a PR instead: https://github.com/filecoin-project/lotus/pull/12361

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -12,10 +12,10 @@
...
   - example: ` fix: mempool: Introduce a cache for valid signatures`
```
[the remark this answers is not in the document]
Comment (BigLep) [reply .github/pull_request_template.md:16]: Superseded by https://github.com/filecoin-project/lotus/pull/12361

File: .github/pull_request_template.md [hunk 6]
```diff
@@ -12,10 +12,10 @@
...
-  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
-  - `area`, e.g. api, chain, state, mempool, multisig, networking, paych, proving, sealing, wallet, deps
+  - `type`: build, chore, ci, docs, feat, fix, perf, refactor, revert, style, test
```
[the remark this answers is not in the document]
Comment (BigLep) [reply .github/pull_request_template.md:17]: Superseded by https://github.com/filecoin-project/lotus/pull/12361

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -12,10 +12,10 @@
...
+  - `scope`: api, chain, deps, mempool, multisig, networking, paych, proving, sealing, state, wallet
```
[the remark this answers is not in the document]
Comment (BigLep) [reply .github/pull_request_template.md:18]: Superseded by https://github.com/filecoin-project/lotus/pull/12361

**Comment by BigLep** [comment]
For the record, concerning using off-the-shelf tools given we have had to iterate on this a couple times (e.g., PR #12373) , I did a little more researching.  I don't think they'll give us fully what we want and we should finish fine-tuning what we started here:

Tool | Has regex support? | Post customizable comment message? | Supports forks (pull_request_target)?
-- | -- | -- | --
https://github.com/marketplace/actions/pr-title-checker | Yes | Doesn't comment.  It applies a label.  No way for us to get a message that links to our conventions. | (Didn't look)
https://github.com/marketplace/actions/pr-title-verify | Yes | Not sure.  Not documented. | Not sure.  Not documented.
https://github.com/Slashgear/action-check-pr-title | Yes | Yes | No

**Pull request merged**